File: tools/parity/dispatch/git-remote-error.ts
// TS dispatch for the git-remote-error parity module. The shared TS reference
// was DELETED (the Rust orca-text core is the sole impl — napi in main, wasm
// in the relay/renderer), so this adapter drives the napi binding: the vectors'
// recorded goldens now pin the napi surface absolutely, and the harness's
// TS-vs-Rust diff degenerates to napi-vs-binary (drift between the two Rust
// entry points would still surface here). Requires the built addon, like the
// napi-parity suite.

import { requireRustGitBinding } from '../../../src/main/daemon/rust-git-addon'

export function dispatch(fn: string, input: unknown): unknown {
  switch (fn) {
    case 'stripCredentialsFromMessage':
      return requireRustGitBinding().stripCredentialsFromMessage(input as string)
    case 'normalizeGitErrorMessage': {
      const { message, operation } = input as {
        message: string | null
        operation?: string
      }
      // A null message models a non-Error throw (Rust `None`); a string models
      // the Error path — mirroring the Rust dispatch's `Option<&str>` mapping.
      return requireRustGitBinding().normalizeGitErrorMessage(message ?? undefined, operation)
    }
    default:
      throw new Error(`unknown function ${fn}`)
  }
}

File: rust/crates/orca-parity/src/modules/git_remote_error.rs
//! Parity dispatch for `orca_text::git_remote_error` vs
//! `src/shared/git-remote-error.ts`.

use orca_text::git_remote_error::{
    normalize_git_error_message, strip_credentials_from_message, GitRemoteOperation,
};
use serde_json::{json, Value};

pub fn dispatch(function: &str, input: &Value) -> Value {
    match function {
        "stripCredentialsFromMessage" => match input.as_str() {
            // TS returns a plain string; mirror it as a JSON string.
            Some(message) => Value::String(strip_credentials_from_message(message)),
            // Vectors only carry string inputs; a non-string is a vector bug.
            None => json!({ "__parity_error__": "expected string input" }),
        },
        "normalizeGitErrorMessage" => {
            // A null/absent message models a non-Error throw (TS `undefined`) ->
            // Option::None; a string models the Error path.
            let message = input.get("message").and_then(Value::as_str);
            let operation = match input.get("operation").and_then(Value::as_str) {
                Some("push") => Some(GitRemoteOperation::Push),
                Some("pull") => Some(GitRemoteOperation::Pull),
                Some("fetch") => Some(GitRemoteOperation::Fetch),
                Some("upstream") => Some(GitRemoteOperation::Upstream),
                _ => None,
            };
            Value::String(normalize_git_error_message(message, operation))
        }
        other => json!({ "__parity_error__": format!("unknown function {other}") }),
    }
}
